diff --git a/src/Makefile b/src/Makefile
index b733a11..289b834 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -135,19 +135,56 @@ blst:
 			$(MAKE) build_blst; \
 	fi
 
+
+# TODO: use shared library for tests instead of this separate compilation!
+
+
 # This compiles the tests with optimizations disabled.
 # It will re-build if any of our source/header files change.
+# tests: CFLAGS += -O0
+# tests: blst $(SOURCE_FILES) $(HEADER_FILES)
+#	@echo "[+] building tests"
+#	@$(CC) $(CFLAGS) -o $@ test/tests.c $(LIBS)
+
+# This simply runs the test suite.
+# .PHONY: test
+# test: tests
+# 	@echo "[+] executing tests"
+#	@./tests
+
 tests: CFLAGS += -O0
-tests: blst $(SOURCE_FILES) $(HEADER_FILES)
+tests: blst libckzg.a
 	@echo "[+] building tests"
-	@$(CC) $(CFLAGS) -o $@ test/tests.c $(LIBS)
+	@$(CC) $(CFLAGS) -o $@ test/tests.c -L$(LIBRARY_OUTPUT) -lckzg $(LIBS)
 
-# This simply runs the test suite.
 .PHONY: test
 test: tests
 	@echo "[+] executing tests"
 	@./tests
 
+# Get all .c files recursively
+SOURCE_FILES := $(shell find . -name '*.c' | sed 's|^\./||' | sort)
+OBJECTS = $(SOURCE_FILES:.c=.o)
+
+# Library output directory
+LIBRARY_OUTPUT = ../lib
+
+# Compilation flags (reusing existing flags from the Makefile)
+CC = clang  # or gcc, depending on platform
+CFLAGS += -I. -I../inc -O2 -Werror -fPIC
+
+$(LIBRARY_OUTPUT):
+	mkdir -p $(LIBRARY_OUTPUT)
+
+# Compile .c files to .o files
+%.o: %.c
+	$(CC) $(CFLAGS) -c $< -o $@
+
+# Build the static library
+libckzg.a: $(OBJECTS) | $(LIBRARY_OUTPUT)
+	ar rcs $(LIBRARY_OUTPUT)/$@ $(OBJECTS)
+
+
 ###############################################################################
 # Coverage
 ###############################################################################
@@ -259,4 +296,5 @@ clean:
 	@echo "[+] cleaning"
 	@rm -f *.o */*.o *.profraw *.profdata *.html xray-log.* *.prof *.pdf \
 	    tests tests_cov tests_prof .blst_hash
+	rm -f ../lib/libckzg.a
 	@rm -rf analysis-report
diff --git a/src/eip4844/eip4844.c b/src/eip4844/eip4844.c
index 9fe2ebb..741125d 100644
--- a/src/eip4844/eip4844.c
+++ b/src/eip4844/eip4844.c
@@ -503,6 +503,19 @@ out:
     return ret;
 }
 
+C_KZG_RET compute_kzg_proof_wrapper(
+    KZGProof *proof_out,
+    fr_t *y_out,
+    const fr_t *polynomial,
+    const size_t size,
+    const fr_t *z,
+    const KZGSettings *s
+) {
+    Polynomial poly;
+    for(size_t i=0;i<size;i++) poly.evals[i] = polynomial[i];
+    return compute_kzg_proof_impl(proof_out, y_out, &poly, z, s);
+}
+
 /**
  * Given a blob and a commitment, return the KZG proof that is used to verify it against the
  * commitment. This function does not verify that the commitment is correct with respect to the
diff --git a/src/eip4844/eip4844.h b/src/eip4844/eip4844.h
index 298da9c..0bbdc27 100644
--- a/src/eip4844/eip4844.h
+++ b/src/eip4844/eip4844.h
@@ -52,6 +52,15 @@ C_KZG_RET compute_blob_kzg_proof(
     KZGProof *out, const Blob *blob, const Bytes48 *commitment_bytes, const KZGSettings *s
 );
 
+C_KZG_RET compute_kzg_proof_wrapper(
+    KZGProof *proof_out,
+    fr_t *y_out,
+    const fr_t *polynomial,
+    const size_t size,
+    const fr_t *z,
+    const KZGSettings *s
+);
+
 C_KZG_RET verify_kzg_proof(
     bool *ok,
     const Bytes48 *commitment_bytes,
